src/midi-player.ts

```diff
@@ -16,31 +17,117 @@ export class MidiPlayer implements IMidiPlayer {
 
     private _offset: null | number;
 
+    private _paused: null | number;
+
+    private _latest: null | number;
+
     private _resolve: null | (() => void);
 
     private _scheduler: Scheduler;
 
     private _schedulerSubscription: null | { unsubscribe(): void };
 
+    private _channels: null | number[];
+
     constructor({ encodeMidiMessage, json, midiFileSlicer, midiOutput, scheduler }: IMidiPlayerOptions) {
         this._encodeMidiMessage = encodeMidiMessage;
         this._endedTracks = null;
         this._json = json;
         this._midiFileSlicer = midiFileSlicer;
         this._midiOutput = midiOutput;
         this._offset = null;
+        this._paused = null;
```

Comment:
I think we can do all necessary computations with only the offset (and without an additional paused property). The offset could be `0` initially.

- `play()` would set the offset to `performance.now()` (minus the previous offset but it's `0`).
- `pause()` would set the offset to the elapsed time.
- `resume()` would set the offset to `performance.now()` minus the previous offset.
- `stop()` would set it back to `0`.

Reply:
Done